src/realm/parser/parser_utils.hpp

```diff
@@ -90,6 +92,32 @@ T stot(std::string const& s) {
     return value;
 }
 
+template <>
+inline double stot(std::string const& s)
+{
+    double value;
+    try {
+        value = stold(s); // parsing NaN and +-Infinity are well defined in this conversion to long double
```

Comment:
why `stold` rather than `stod`? do you really want `long double` support? Also I'm pretty sure you don't want to use the `stoX` family of functions here:

* they don't throw exceptions and report errors using and optional second parameter
* they are locale sensitive, so whether the decimal point is `.` or `,` depends on the user's settings
* they accept hex floats, which I assume we don't want to be part of the query language (unless we do?)
* they allow arbitrary garbage at the end of the string (eg `"1 singular sensation"` is parsed as `1.0`)
* they support NaN-boxed values using a platform-specific format
* We've had [issues with parsing nan/inf on windows](https://github.com/mongodb/mongo/blob/5d376083d4c5e597bda2b967cfcce32874025b12/src/mongo/base/parse_number.cpp#L254-L274) (maybe this has been fixed in newer versions?)

https://en.cppreference.com/w/cpp/string/basic_string/stof

Unfortunately there aren't any great options here. [`from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars) from c++17 is *supposed* to be the right answer, but it still isn't implemented by many stdlibs (it has been called the "final boss" of c++17 by a few implementers). As much as I am loathe to suggest using iostreams, I think the best option is to use an `istringstream` with `.imbue(std::locale::classic())` to at least force use of the plain "C" locale so it is at least locale independent, and error handling is via exceptions. It doesn't solve all the problems, but it at least fixes the worst of them. We could do some extra checks to ensure that the input is "sane" if we want to ensure that it is only using normal floating point numbers.

Reply:
Update: I asked about it on the cpplang slack and someone recomended https://github.com/ulfjack/ryu. I think it is the considered the best in class algorithm (fairly recently discovered and multiple orders of magnitude better than prior). If we can pull in a 3rd party lib, we should strongly consider pulling this in. It is what MSVC is using in their `from_chars` implementation

Reply:
How did we convert from string to float/double before? I guess we have always been locale dependend. So I think the aim of this PR is to add support for exponential notation - not to solve all problems in the world. I you get what I mean.

Reply:
> How did we convert from string to float/double before?

As far as I can tell, we didn't. We only accepted integers, even when querying against float/double columns: https://github.com/realm/realm-core/blob/core-6.1/src/realm/parser/value_expression.cpp#L155-L171. I couldn't find any of our code that parses to floats, but maybe I wasn't grepping for the right thing. 

I did stumble across this example in the pegtl codebase for parsing a double: https://github.com/taocpp/PEGTL/blob/5ca0db3b0e507b410c89eb28af99f8e3b73f18cf/src/example/pegtl/double.hpp. Well, it only handles recognizing a float, but leaves the actual task of making the corresponding double as an exercise to the reader, so perhaps it isn't so useful.

Reply:
The generic `stot` function above actually does work for floats and doubles, and even handles exponentials correctly. However, for some reason, parsing +-`infinity` and `NaN` does not work on linux/android (I was surprised to see this because all tests passed with the generic function on my mac). While the ryu library looks really nice, I'd probably prefer to use the generic istringstream implementation, and if it fails to parse, check for infinity and NaN manually.

Reply:
The generic one is ok. It can be improved with `iss.imbue(std::locale::classic())`, but I agree that it is reasonable to defer fixing that outside of this review if you would prefer, since it was an existing issue. Adding special case checks for inf and nan sounds fine to me.

Reply:
Updated this to use `iss.imbue(std::locale::classic())`, thanks for pointing that out. Other locales wouldn't actually work, because the parser syntax is hard coded to only support `.` so any locale that uses `,` as a decimal separator wouldn't even make it this far because the parser would mark it as invalid syntax.
I also ripped off the mongo parsing code you linked to because that's basically what we needed here. I had to add a few more cases though: `-inf`, `inf`, `+inf`, `-nan`, `+nan` is it a bug that mongo's windows implementation doesn't support these cases?